[MRG] add group information to Searchlight

njchiang opened the pull request:
-updates sklearn.cross_validation to sklearn.model_selection: cross_val_score() now takes a groups argument before scoring.
-adds group label argument (default None) for cross validation schemes accepting labels such as LeaveOneGroupOut(). This might be useful for those interested in leaving one run out at a time in their cross-validation schemes. 

example: 
cv = model_selection.LeaveOneGroupOut()
sl = decoding.Searchlight(cv=cv)
sl.fit(imgs, y, groups)

bthirion commented:
Sounds good, thanks !
One thing that is missing now is some units tests to check that the groups variable is used correctly.

njchiang commented:
No problem!

I believe cross_val_score() raises an error when lengths of X, y, and groups are different, e.g.
ValueError: Found input variables with inconsistent numbers of samples: [288, 288, 864]. 
It is the same error that it raises when the data (imgs) and target (y) are of unequal lengths.

Should I introduce a test condition into test_searchlight.py?

I should note that by using model_selection, the cross_validation will no longer be compatible with scikit-learn < 0.15 (I think that is when model_selection was introduced)

KamalakerDadi commented:
>I should note that by using model_selection, the cross_validation will no longer be compatible with scikit-learn < 0.15 (I think that is when model_selection was introduced)

You can have a look at ```nilearn/_utils/fixes/__init__.py```.

njchiang commented:
I hope I haven't broken anything. I've added two tests to test_searchlight.py, one which checks proper usage and one which checks if someone unnecessarily specified the groups. 

I also tried to increase compatibility in searchlight.py by moving "groups" into kwargs, which firsts checks that sklearn is >0.15. There is one more issue in that I modified the import statement to be from model_selection and not cross_validation (which is the case in <0.15) and will look into a workaround.

KamalakerDadi commented on nilearn/_utils/fixes/__init__.py:
```diff
@@ -5,12 +5,15 @@
 except ImportError:
     # scikit-learn < 0.16
     from sklearn.cross_validation import _check_cv as check_cv
+    from sklearn.cross_validation import cross_val_score
```
I don't think this is required.

KamalakerDadi commented:
Thanks @njchiang. Overall your PR looks good. I will have a closer look in mean time.

Minor comment: Could you address flake8 errors. Basically there are too long line written
https://travis-ci.org/nilearn/nilearn/jobs/258020225

KamalakerDadi commented on nilearn/decoding/searchlight.py:
```diff
@@ -50,6 +50,9 @@ def search_light(X, y, estimator, A, scoring=None, cv=None, n_jobs=-1,
         adjacency matrix. Defines for each feature the neigbhoring features
         following a given structure of the data.
 
+    groups : array-like
```
You can take optional beside array-like to short the line.

array-like, optional

KamalakerDadi commented on nilearn/decoding/searchlight.py:
```diff
@@ -50,6 +50,9 @@ def search_light(X, y, estimator, A, scoring=None, cv=None, n_jobs=-1,
         following a given structure of the data.
 
+    groups : array-like, optional
+        group membership for each sample for cross validation. default None
```
I would write membership -> labels to be more precise for neuroimaging users.

KamalakerDadi replied:
I think we should write explicitly that groups is only available for scikit-learn version >= 0.18

May be we should add as a NOTE below default None.

NOTE: this argument is valid for scikit learn version >= 0.18 ?

Does this sounds good ? ping @GaelVaroquaux @bthirion

KamalakerDadi commented on nilearn/decoding/searchlight.py:
```diff
@@ -130,6 +133,9 @@ def _group_iter_search_light(list_rows, estimator, X, y,
         target variable to predict.
 
+    groups : array-like, optional
+        group membership for each sample for cross validation.
```
same here membership -> labels

KamalakerDadi commented on nilearn/decoding/searchlight.py:
```diff
@@ -273,6 +281,10 @@ def fit(self, imgs, y):
             3D images in img.
 
+        groups : array-like, optional
+            group membership for each sample for cross validation. Must have
```
same here membership -> labels

KamalakerDadi commented on nilearn/decoding/searchlight.py:
```diff
@@ -273,6 +281,10 @@ def fit(self, imgs, y):
+        groups : array-like, optional
+            group membership for each sample for cross validation. Must have
+            exactly as many elements as 3D images in img. default None
+
```
same here adding NOTE to say users that this argument only available for scikit learn >= 0.18

KamalakerDadi commented on nilearn/decoding/tests/test_searchlight.py:
```diff
@@ -72,3 +72,29 @@ def test_searchlight():
+    assert_equal(np.where(sl.scores_ == 1)[0].size, 7)
+    assert_equal(sl.scores_[2, 2, 2], 1.)
+
+    # adding superfluous group variable
```
What's the difference in this test block compared to above one ?
I see no difference. They are the same.

njchiang replied:
The results should be identical to the test in lines 57-67. The only difference is in sklearn >= 0.18 this will try LeaveOneGroupOut (cv=gcv) in line 88 rather than the KFold declared above. Let me know if I'm misunderstanding.

KamalakerDadi commented on nilearn/decoding/tests/test_searchlight.py:
```diff
@@ -72,3 +72,29 @@ def test_searchlight():
+
+    # adding superfluous group variable
+    sl = searchlight.SearchLight(mask_img, process_mask_img=mask_img, radius=1,
+                                 n_jobs=n_jobs, scoring='accuracy', cv=cv)
```
May be you should add tests with other cv ?
For instance, KFold to check that groups doesn't fail with other cvs

njchiang replied:
yes, this test uses KFold (cv=cv), which does not take a groups argument (even though I pass one in line 96). Is this what you mean? I can add other conditions you have in mind.

KamalakerDadi replied:
Yes, you are right. Sorry, you can ignore this comment.

KamalakerDadi commented on nilearn/decoding/searchlight.py:
```diff
@@ -50,6 +50,9 @@ def search_light(X, y, estimator, A, scoring=None, cv=None, n_jobs=-1,
         following a given structure of the data.
 
+    groups : array-like, optional
+        group label for each sample for cross validation. default None
```
Can you please add NOTE here too ?

This function is also made available to users same like class.

KamalakerDadi commented:
Apart from my small comment. I am +1 to merge.

bthirion commented:
LGTM overall. Thx.

KamalakerDadi commented:
Merged. Thank you for your work.

njchiang commented:
Thank you!
